numpy/lib/function_base.py

```diff
@@ -3612,17 +3786,81 @@ def _quantile_is_valid(q):
     return True
 
 
-def _quantile_ureduce_func(a, q, axis=None, out=None, overwrite_input=False,
-                           interpolation='linear', keepdims=False):
+def _validate_weights(a, q, axis, weights):
+    if weights is None:
+        wgt = None
+    else:
+        a = asanyarray(a)
+        wgt = np.asanyarray(weights)
+
+        if issubclass(a.dtype.type, (np.integer, np.bool_)):
+            result_dtype = np.result_type(a.dtype, wgt.dtype, 'f8')
+        else:
+            result_dtype = np.result_type(a.dtype, wgt.dtype)
+
+        broadcastable = False
+        if a.ndim == wgt.ndim and a.shape != wgt.shape:
+            broadcastable = all([a_dim == w_dim or w_dim == 1
+                                 for a_dim, w_dim
+                                 in zip(a.shape, wgt.shape)])
+
+        if a.shape != wgt.shape and not broadcastable:
+            if axis is None:
+                raise TypeError(
+                    "Axis must be specified when shapes of a and weights "
+                    "differ and not broadcastable.")
+            if wgt.ndim != 1:
+                raise TypeError(
+                    "1D weights expected when shapes of a and weights differ "
+                    " and not broadcastable.")
+            if wgt.shape[0] != a.shape[axis]:
+                raise ValueError(
+                    "Length of weights not compatible with specified axis.")
+            if not np.issubdtype(wgt.dtype, np.number):
+                raise ValueError("All weight entries must be numeric.")
+
+            nan_ws = wgt[np.isnan(wgt)]
+            if nan_ws.size > 0:
+                raise ValueError("No weight can be NaN.")
+
+            negative_ws = wgt[wgt < 0]
+            if negative_ws.size > 0:
+                raise ValueError("Negative weight not allowed.")
+
+            # setup wgt to broadcast along axis
+            wgt = np.broadcast_to(wgt, (a.ndim-1)*(1,) + wgt.shape)
+            wgt = wgt.swapaxes(-1, axis)
+        else:  # same shape, or at least broadcastable
+            if axis is None:
+                axis = tuple(range(a.ndim))
```

Comment:
Is this necessary? Doesn't `_ureduce` handle it?

Reply:
I explicitly set this to avoid the `ap = a.flatten()` statement in `_quantile_reduce_func`.  I've already set up `wgt` to match the dimensions of `a`, so the flattening messes things up.